app/webpacker/controllers/login_modal_controller.js

```diff
@@ -1,8 +1,74 @@
 import { Controller } from "stimulus"
+import CableReady from "cable_ready"
 
 export default class extends Controller {
+  static targets = ["background", "modal", "email"]
+  static values = { email: String }
+
+  connect() {
+    if(this.hasModalTarget) {
+      window.addEventListener("login:modal:open", this.open)
+
+      if(location.hash.substr(1).includes("/login")) {
+        this.open()
+      }
+    }
+  }
+
   call(event) {
     event.preventDefault()
     window.dispatchEvent(new Event("login:modal:open"))
   }
+
+  emailOnInput(event) {
+    this.emailValue = event.currentTarget.value
+    this.emailTargets.forEach((element) => {
+      element.value = this.emailValue
+    })
+  }
+
+  open = () => {
+    if(!location.hash.substr(1).includes("/login")) {
+      history.pushState({}, "", "#/login")
+    }
+
+    this.backgroundTarget.style.display = "block"
+    this.modalTarget.style.display = "block"
+
+    document.querySelector("body").classList.add("modal-open")
+
+    setTimeout(() => {
```

Comment:
why this needs a `setTimeout()` ?

Reply:
So what's happening here is we're doing a show/hide type thing with the modal. But there's a slight animation on it appearing/disappearing (which is partly controlled by CSS animations via the `fade` and `in` classes). If we just toggle `display: block` or `display: none` it's **instantaneous** and doesn't give any time for the animation to be seen. So instead we toggle the class which triggers the animation, wait 200 milliseconds, and then toggle display block/none.

It feels a little bit convoluted, but I was just replicating the existing behaviour :man_shrugging:

Reply:
I don't see any `delay` here (although I can see it for the `close` method).
You mean:
```js
setTimeout(() => {
  this.modalTarget.classList.add("in")
  this.backgroundTarget.classList.add("in")
}, 200)
```

Reply:
Ah right! I missed the delay here, it should be on both.

Nice catch :pray:

Reply:
Great! I thought maybe you wanted to not be in the same "thread" of execution (even if JS is single thread) for any reason... But that's simpler, only a forgot ;)

Reply:
Hmmm I just had another look and it's a bit weird: if I remove the `setTimeout` completely there's no animation, but if I use `setTimeout` with no delay (0 milliseconds) it works fine. :thinking:

Reply:
Yes, probably because it needs to be executed not in the same context as the rest of the method (why, I don't know to be honest)